playground/src/app.jsx

```diff
@@ -488,6 +509,67 @@ function App() {
     }
   }, [monaco]);
 
+  React.useEffect(() => {
+    let hoverProvider = undefined;
+    if (monaco) {
+      hoverProvider = monaco.languages.registerHoverProvider(language, {
+        provideHover: function (model, position) {
+          const { lineNumber, column } = position;
+          if (!compilation?.typeHints) {
+            return null;
+          } else {
+            function hintInLoc(item) {
+              var end = item.end;
+              var start = item.start;
+              const result =
+                lineNumber >= start.line &&
+                lineNumber <= end.line &&
+                column >= start.col + 1 &&
+                column <= end.col + 1;
+              return result;
+            };
+            const result = compilation?.typeHints.find(hintInLoc);
+            if (result) {
+              const range = new monaco.Range(
+                result.start.line,
+                result.start.col + 1,
+                result.end.line,
+                result.end.col + 1
+              );
+              let hint = result.hint;
+              if (language == languageMap.Reason) {
+                const prefix = "type t = ";
+                try {
+                  hint =
+                    ocaml
+                    /* add prefix so it is valid code */
+                    .printRE(ocaml.parseML(prefix + hint))
```

Comment:
is there any case where this can be problematic?

Reply:
where it can raise an exn? who knows... 🤷  just playing it safe.

Reply:
I mean the prefix `type = t` being added to the hint. Do recursive types might break here? or other cases?

Reply:
Ah, good point! After your comment, I tested more thoroughly and found out this approach would break for type definitions, where we already have a valid AST and we can parse directly. I fixed it in https://github.com/melange-re/melange-re.github.io/pull/122/commits/fa07c2962c2baf6824ecef33e669768b3846d318.